src/NonCumulative/Soundness/Realizability.agda

```diff
@@ -0,0 +1,527 @@
+              helper {Δ = Δ} ⊢σ with W′ , ↘W′ , _ ← 𝕌⊆TopT A∈𝕌 (len Δ)
+                                   | W , ↘W , W≈ ← rc (≤-reflexive refl) A∈𝕌 rel ⊢σ -- well-founded induction
+                                  rewrite Rty-det ↘W ↘W′ = ≈-conv W≈ (≈-trans (≈-sym (Se-[] _ (⊢w⇒⊢s ⊢σ))) ([]-cong-Se′ (≈-sym T≈) (⊢w⇒⊢s ⊢σ)))
+      ®El⇒®↑El {Π j A (S ↙ k) ρ} {Γ = Γ} {t = t} {T = T} rc (Π′ {j} {k} jA RT) t® = record
+        { t∶T = t∶T
+        ; T∼A = ®El⇒® (Π′ jA RT) t®
+        ; a∈⊤ = El⊆Top (Π′ jA RT) a∈El
+        ; krip = λ {Δ} {σ} ⊢σ →
+                 let W , ↘W , _  =  El⊆Top (Π′ jA RT) a∈El (len Δ)
+                     in subst (_ ⊢ _ ≈_∶[ _ ] _) (cong Nf⇒Exp (Rf-det ↘W (proj₁ (proj₂ (El⊆Top (Π′ jA RT) a∈El (len Δ))))))
+                        (helper ⊢σ ⊢IT ⊢OT jA RT krip  ↘W )
+        }
+        where open GluΛ t®
+              helper :  Δ ⊢w σ ∶ Γ →
+                        Γ ⊢ IT ∶[ ℕ.suc j ] Se j →
+                        (IT ↙ j) L.∷ Γ ⊢ OT ∶[ ℕ.suc k ] Se k →
+                        (jA : A ≈ A′ ∈ PERDef.𝕌 j (λ l<j → 𝕌-wellfounded (max j k) (ΠI≤ refl l<j))) →
+                        (RT : ∀ {a a′} → a ≈ a′ ∈ PERDef.El j _ jA → ΠRT S (ρ ↦ a) T′ (ρ′ ↦ a′) (PERDef.𝕌 k (λ l<k → 𝕌-wellfounded (max j k) (ΠO≤ refl l<k))) ) →
+                        (∀ {Δ σ} → Δ ⊢w σ ∶ Γ → ΛRel j Δ t IT OT σ a (λ l<j → 𝕌-wellfounded (max j k) (ΠI≤ refl l<j)) jA
+                          (Glu.⟦ j , (λ l<j → Glu-wellfounded (max j k) (ΠI≤ refl l<j)) , _ ⟧_⊢_® jA)
+                          (Glu.⟦ j , (λ l<j → Glu-wellfounded (max j k) (ΠI≤ refl l<j)) , _ ⟧_⊢_∶_®_∈El jA)
+                          (λ a∈ Δ′ s S b → Glu.⟦ k , (λ l<k → Glu-wellfounded (max j k) (ΠO≤ refl l<k)) , (λ l<k → 𝕌-wellfounded (max j k) (ΠO≤ refl l<k)) ⟧ Δ′ ⊢ s ∶ S ® b ∈El ΠRT.T≈T′ (RT a∈))) →
+                        Rf L.foldr (λ _ → ℕ.suc) 0 Δ - ↓ (max j k) (Π j A (S ↙ k) ρ) a ↘ W →
+                        --------------------------
+                        Δ ⊢ sub t σ ≈ Nf⇒Exp W ∶[ max j k ] sub T σ
+              helper {Δ = Δ} {σ = σ} {W = Λ (W ↙ _) w} ⊢σ ⊢IT ⊢OT jA RT krip (RΛ .(len Δ) ↘W ↘a ↘⟦S⟧ ↘w _)
+                rewrite Glu-wf-gen j (ΠI≤′ j k refl)
+                      | Glu-wf-gen k (ΠO≤′ j k refl)
+                      | 𝕌-wf-gen j (ΠI≤′ j k refl)
+                      | 𝕌-wf-gen k (ΠO≤′ j k refl) = helper₁
+                where ⊢σ′   = ⊢w⇒⊢s ⊢σ
+                      Tσ≈   = ≈-trans ([]-cong-Se′ T≈ ⊢σ′) (Π-[] ⊢σ′ ⊢IT ⊢OT refl)
+                      ⊢tσ   = conv (t[σ] t∶T ⊢σ′) Tσ≈
+                      ⊢Δ    = proj₁ (presup-s ⊢σ′)
+                      ⊢Γ    = proj₂ (presup-s ⊢σ′)
+                      ⊢ITσ  = t[σ]-Se ⊢IT ⊢σ′
+                      ⊢ITσΔ = ⊢∷ ⊢Δ (t[σ]-Se ⊢IT ⊢σ′)
+                      ⊢qσ   = ⊢q ⊢Δ ⊢σ′ ⊢IT
+                      ⊢OTqσ = t[σ]-Se ⊢OT ⊢qσ
+                      ⊢σwk  = s-∘ (s-wk ⊢ITσΔ) ⊢σ′
+                      open ΛRel (krip ⊢σ) using (IT-rel)
```

Comment:
I would just assign different names to the module and use `.` operator. this might gets tricky after some time.